### ISSUE TITLE
Topology simulation

### ISSUE THREAD

**Issue by a28283878** [nyhuang/HQTimer#3]
There are plenty of topology in your code, could you tell me which one did you use to simulate your journal.

**Comment by nyhuang** [comment]
setting.Single and setting.Bridge are used. Some more topologies have been tested by my collaborators, but they are not recorded in the code or the paper.

**Comment by a28283878** [comment]
![image](https://user-images.githubusercontent.com/16368909/90413339-e7887f80-e0e0-11ea-937b-156f8371af61.png)
may I ask which topologies did you use for this graph.
And does the dataset of hard and hybrid are the same? the difference of overflow number and total install are super big.

**Comment by nyhuang** [comment]
setting.Single. They are the same. IMHO, because of the Rule Dependency Problem, the Hard installs too many rules too frequently.

**Comment by a28283878** [comment]
Thanks !!
by the way, I see you use 'single_rule_0.2.pkl' as the rule pkl of Single simulation.
So you first use real traffic pcap to transform into 'single.pkl', and then you use 'single.pkl' to generate the rule with 10~90% of wildcard rule. And you use the 20% wildcard rule, 'single_rule_0.2.pkl', as the ruleset.
Am I correct?

**Comment by nyhuang** [comment]
Yes I think so... Sorry it has been a long time therefore I cannot recall all the details. But your descriptions sound right.

**Comment by a28283878** [comment]
Okay, Thanks for your help. I appreciate your help very much.
By the way, I use the trace log from MAWI 2009/12/18. I slice the trace log into 100K flows, and run the simulation.

I found out that in hybrid mode. Overflow number of ITM increase sharply in the first 7000 flows, but stop increasing in the rest of the flow and overflow number stay around 9000 times.
Q or DQN increase slowly in the first 7000 flow, but continue to increase in the rest of the flow and exceed ITM after 8000th flow.

Do you remember any situation like this when you simulated?

**Comment by nyhuang** [comment]
No. Sounds interesting.

IMHO, Q and DQN are not easy to train. The training quality depends highly on the dataset. Maybe you can change a dataset and see if the results look alike. You can also save the model and observe its behavior. As for the ITM, no idea why it behaves like you said.

The training process of DQN can be improved in many ways: the input should be normalized, the network can be pre-trained, etc. I am not an expert on DRL or DNN. Yann LeCun has published a paper _Efficient BackProp_ and introduced some very basic techniques.

**Comment by a28283878** [comment]
I did some research about the result.

Exactly like you mentioned. ITM start to set timeout as 1s since 2200th flow,.But DQN and Q seem to be untrained and choosing timeout randomly.

So before you simulate, would you use DQN and Q to train a model and use that model to simulate to get a better result.
Or you use DQN and Q to run simulation and train at the same time.

Because the journal looks like you train your model every 1 second in simulation. Instead of using a well-trained model to run  the simulation.

**Comment by nyhuang** [comment]
I use DQN and Q to run simulation and train at the same time.

**Comment by a28283878** [comment]
Okay, I will try more.
 Thanks for your help.
